keccak256/src/circuit.rs

```diff
@@ -0,0 +1,284 @@
+    xi_config: XiConfig<F>,
+    iota_b9_config: IotaB9Config<F>,
+    iota_b13_config: IotaB13Config<F>,
+    absorb_config: AbsorbConfig<F>,
+    state: [Column<Advice>; 25],
+    _marker: PhantomData<F>,
+}
+
+impl<F: FieldExt> KeccakFConfig<F> {
+    // We assume state is recieved in base-9.
+    pub fn configure(
+        &self,
+        meta: &mut ConstraintSystem<F>,
+    ) -> KeccakFConfig<F> {
+        let state = (0..25)
+            .map(|_| {
+                let column = meta.advice_column();
+                meta.enable_equality(column.into());
+                column
+            })
+            .collect_vec()
+            .try_into()
+            .unwrap();
+
+        // TMP for mixing
+        let next_input = (0..ABSORB_NEXT_INPUTS)
+            .map(|_| meta.advice_column())
+            .collect_vec()
+            .try_into()
+            .unwrap();
+        // theta
+        let theta_config = ThetaConfig::configure(meta.selector(), meta, state);
+        // rho
+        let rho_config = RhoConfig::configure(meta.selector(), meta, state);
+        // Pi
+        let pi_config = PiConfig::configure(meta.selector(), meta, state);
+        // xi
+        let xi_config = XiConfig::configure(meta.selector(), meta, state);
+        // Iotab9
+
+        let iota_b9_config = {
+            // Generate advice and instance column for Round constants in base9
+            let round_ctant_b9 = meta.advice_column();
+            let round_constants_b9 = meta.instance_column();
+            IotaB9Config::configure(
+                meta.selector(),
+                meta,
+                state,
+                round_ctant_b9,
+                round_constants_b9,
+            )
+        };
+
+        // Just TMP until Mixing stage is done
+        let iota_b13_config = {
+            // Generate advice and instance column for Round constants in base13
+            let round_ctant_b13 = meta.advice_column();
+            let round_constants_b13 = meta.instance_column();
+            IotaB13Config::configure(
+                meta.selector(),
+                meta,
+                state,
+                round_ctant_b13,
+                round_constants_b13,
+            )
+        };
+
+        // Absorb TMP
+        let absorb_config =
+            AbsorbConfig::configure(meta.selector(), meta, state, next_input);
+
+        KeccakFConfig {
+            theta_config,
+            rho_config,
+            pi_config,
+            xi_config,
+            iota_b9_config,
+            iota_b13_config,
+            absorb_config,
+            state,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn assign_state(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        state: [F; 25],
+    ) -> Result<[F; 25], Error> {
+        // In case is needed
+        let mut state = state;
+        let mut offset = offset;
+
+        // First 23 rounds
+        for round in 0..24 {
```

Comment:
This loop has 24 rounds, where we want only 23.

We can get the constant 24 from `PERMUTATION` in `common.rs`

Reply:
Makes sense. Will edit that as soon as I'm done with the `MixingConfig` and I start polishing the general `KeccakFConfig`.

Good catch BTW!!